Add self-update for West and the manifest + CI fix for Python 3.4

ulfalizer opened the pull request:
CI fix:

```
Add setuptools-scm to test_requirements.txt to work around an issue on
Python 3.4:

  distutils.errors.DistutilsError: Could not find suitable
  distribution for Requirement.parse('setuptools-scm>=1.15.0')

Not sure why it didn't appear before. Maybe something in CI changed.

See pypa/pypi-legacy#322 and
pytest-dev/pytest-xdist#136. I suspect the
issue is related to setup_requires= using easy_install.
```

Main change:

```
Add self-update for West and the manifest

The self-update mechanism checks the West and manifest repositories in
west/ for upstream changes and rebases them to the latest version if any
are found.

'west fetch' and 'west pull' run a self-update before fetching any
project data. It probably makes sense to always work with the latest
version of West and the manifest there. The self-update can be
suppressed with --no-update.

There's also a 'west update' command for manually updating West and/or
the manifest.

West automatically restarts itself after a self-update, before running
any commands or parsing the manifest. This makes it safe to add a new
feature to West and immediately start using it in the manifest.

The restarting is done with a plain exec(3). Getting this to work
required a few changes:

 - West now uses absolute instead of relative imports, to make it
   possible to execute src/west/main.py directly. This also makes it a
   bit simpler to work on West. Relative imports seem to be discouraged
   when googling around a bit.

   __init__.py was removed from src/west. West is more of an application
   than a package now (though it uses some packages internally).

 - The internal 'cmd' package was renamed to 'commands', to avoid
   clashing with a module in the Python standard library.

   'runner' was renamed to 'runners' as well, for consistency.

The tests now have to be run from the West repository root with
'PYTHONPATH=src/west pytest'. The previous testing setup "just working"
relied on having run 'pip3 install -e .' first, which happened to add
src/west to the path due to the bootstrap script being there.

__init__.py was removed from the test directories as well. Packages
mostly make sense for tests if there are naming collisions, and naming
collisions will be obvious if packages aren't used.
```

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -604,6 +671,69 @@ def _checkout(project, branch):
+        merge_base = attempt(project, 'merge-base HEAD remotes/origin/master')
+
+        # Get the current SHA of the upstream branch
+        head_sha = attempt(project, 'show-ref --hash remotes/origin/master')
```
Upstream branch hardcoded for now.

ulfalizer replied:
The logic here could be used for normal projects later as well, to check if they are up-to-date.

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -604,6 +671,69 @@ def _checkout(project, branch):
+        attempt(project, 'fetch')
+
+        # Get the SHA of the last commit in common with the upstream branch
+        merge_base = attempt(project, 'merge-base HEAD remotes/origin/master')
```
Upstream branch hardcoded for now.

ulfalizer replied:
Think it would need the configuration file, so better to do it later.

ulfalizer commented:
recheck

ulfalizer commented:
Cleaned things up a bit and added a precaution to the restart-after-update code: It now throws an exception which is caught in `main.py`, and the `exec()` is done from there.

That makes sure that file buffers in any surrounding `with open(...)`, etc., blocks are flushed, though there aren't any at the moment.

Should be ready now.

carlescufi commented:
@ulfalizer thanks for the PR. Reading through this I have some initial thoughts after picturing the workflow in my head.

>west fetch and west pull run a self-update before fetching any
project data. It probably makes sense to always work with the latest
version of West and the manifest there. The self-update can be
suppressed with --no-selfupdate.

So  the first thing that comes into my mind is whether we really want self-udpate to be performed automatically every time we fetch. A bit like `apt update` vs `apt upgrade`, though different in nature, I wonder if we should have `west update` and then `west fetch` or similar. Just fishing for comments here, I have no definite opinion no this, what are your thoughts?

>There's also a west selfupdate command for manually updating West
and/or the manifest.

Right, this is great because I am sure some conservative users will not want to update west itself in some cases, or the manifest.

ulfalizer commented:
> So the first thing that comes into my mind is whether we really want self-udpate to be performed automatically every time we fetch. A bit like `apt update` vs `apt upgrade`, though different in nature, I wonder if we should have `west update` and then `west fetch` or similar. Just fishing for comments here, I have no definite opinion no this, what are your thoughts?

This was modeled after `repo`, which self-updates whenever you `sync`.

One reason I think it makes sense is that you'd probably want the latest manifest version whenever you fetch/pull new project new data. Since the latest manifest version might be using new West features, that indirectly means that you also have to update West, in order to be sure that you can parse the new manifest.

ulfalizer commented:
Maybe auto-sync could be turned into an option. I suspect we might need a configuration file at some point anyway (something like `git config`), e.g. to remember the upstream branch for the West manifest.

carlescufi commented:
@ulfalizer thanks for the feedback. I agree with you, you make good points and I'm sold on the idea.

>Maybe auto-sync could be turned into an option. I suspect we might need a configuration file at some point anyway (something like git config), e.g. to remember the upstream branch for the West manifest.

I was just going to suggest this. A `~/.westconfig` would be perfect, but we need to think how to implement that on Windows. I think `%userprofile%\westconfig` (hidden) is probably the way to go.

ulfalizer commented on .shippable.yml:
```diff
@@ -13,5 +13,5 @@ build:
     - mkdir -p shippable/{testresults,codecoverage}
-    - pytest --junitxml=shippable/testresults/nosetests.xml tests
-    - pytest --cov=west --cov-report=xml:shippable/codecoverage/coverage.xml tests
+    - PYTHONPATH=src/west pytest --junitxml=shippable/testresults/nosetests.xml tests
```
Got any suggestions for how to fix it? Didn't know we had a thing for running tests through `setup.py`.

mbolivar replied:
I guess the setup.cfg alias entry will need some changes? I'm not sure exactly why this patch set breaks it, is it the elimination of relative imports?

ulfalizer replied:
> I actually think it'd be nice to try to have per-installation configuration to start with -- so `TOP/west/westconfig.yml` -- thoughts?

I'd prefer some simple `.ini`-like format, like Git uses. Think `.yml` might be overkilling it.

ulfalizer replied:
This seems nice: https://docs.python.org/3/library/configparser.html

ulfalizer replied:
I'm not sure why it worked before to be honest. I don't think `pytest` can guess where the code-to-be-tested is, and in this case it's two directories down as well.

Google suggests setting `PYTHONPATH`. Did you put in some helper for `pytest` to localize things somewhere?

ulfalizer replied:
(`PYTHONPATH=src/west python3 setup.py test` works by the way.)

ulfalizer replied:
Looking at https://github.com/pytest-dev/pytest/issues/2421, it seems that `PYTHONPATH` might be needed (they recommend testing against installed packages too).

Still no idea how it worked before. Maybe it just happened to work through the test directories being laid out the same way and some magic relative import thing happening, or maybe I'm still missing something...

ulfalizer replied:
I updated the instructions for running the tests in the README.

I think `pytest-runner` assumes that `setup.py` will install everything that needs to be tested (so that it ends up in `sys.path`). That causes issues in our case, since only the bootstrap script gets installed.

ulfalizer replied:
Finally figured out what's going on:

When the bootstrap script is installed with `pip install -e` (as opposed to a plain `pip install`), the directory `west/src` ends up in `sys.path`, since the `bootstrap` package happens to be in there.

That indirectly makes `west/src/west` importable, provided it is a module, which is how the tests found them.

IOW, `pytest` previously relied on having run `pip install -e .` first.

ulfalizer commented on src/west/commands/build.py:
```diff
@@ -5,11 +5,11 @@
-from . import WestCommand
+import log
+import cmake
+
```
I like putting a blank line before the `from` imports. I could change it.

ulfalizer replied:
Fixed

ulfalizer commented on src/bootstrap/main.py:
```diff
@@ -219,35 +213,42 @@ def init_reinit(directory, args):
+    # sys.argv[1:] strips the argv[0] of the wrapper script itself
+    argv = [sys.executable,
+            os.path.join(topdir, WEST, 'west', 'src', 'west', 'main.py')] \
+           + sys.argv[1:]
```
I was never a fan of that recommendation, so I ignore it. When you see the `(`, you don't know whether some intermediate result is being calculated (think `(a + b)/2`), or if someone just really hates `\`. The `\` makes things clearer.

Are we doing strict PEP 8?

mbolivar replied:
Guido hates `\` IIRC.

We had discussed coming up with a flake8 configuration file at some point in the past but never came up with a solution, so I thought the idea was just to keep things clean based on what it outputs at default settings; did I forget something?

ulfalizer replied:
I'd say just skip the tool and use our common sense.

mbolivar replied:
I actually disagree. Keeping the flake8 output clean may be a bit of extra work, but it means that any time it complains, we can be sure it's based on just the changes introduced by the patchset we are working on. Which includes real bugs sometimes (e.g. flake8 complained that platform was undefined in this patchset when it ran as part of my editor's default config for opening a python file).

ulfalizer replied:
I suspect `flake8` might be confused in this case. It has no problem with this code, for example:

```python3
def f(x):
    return 1 \
           + 2
```

In general, `flake8` doesn't seem to care whether you use `\` or brackets for line continuations. Probably because PEP 8 is a bit wishy-washy on it ("is preferred").

Looking [at the warning itself](https://lintlyci.github.io/Flake8Rules/rules/E127.html), it doesn't make any sense in this case. What it's saying is "hey, I think this code tries to visually align stuff, but it isn't correctly aligned". It's not saying "visual indent is bad". I misread it like that too.

I think it's a bad idea to change stuff just because some tool says so, unless you can think of a way in which it's bad. At that point, it doesn't matter if it was pointed out by a tool or not.

`flake8` and `pylint` (the more OCD one) and the like are nice helpers. Don't take them as dogma.

Breaking rules often makes sense.

mbolivar replied:
Would you be open to using `noqa: xxx` comments on each line that flake8 complains about?

It catches real bugs often enough that cluttering up its output seems like a mistake.

ulfalizer replied:
Not a fan of cluttering up code with stuff like that in general. I'm more of a remove-everything guy. Might've noticed. :P

Would you be fine with just having `flake8` report a few warnings, or suppressing them in a configuration file?

ulfalizer replied:
`flake8` found a single unused variable in Kconfiglib at least. That's still a "run occassionally to check for issues" to me.

I think most of that stuff can be avoided just by working hard to keep code minimalistic and boring, which is why I avoid clutter (except I probably over-comment stuff sometimes...).

ulfalizer commented on src/bootstrap/main.py:
```diff
@@ -219,35 +213,42 @@ def init_reinit(directory, args):
+
+    # sys.argv[1:] strips the argv[0] of the wrapper script itself
+    argv = [sys.executable,
+            os.path.join(topdir, WEST, 'west', 'src', 'west', 'main.py')] \
```
Ah, sorry, had missed those other ones.

ulfalizer replied:
Fixed

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -330,11 +373,34 @@ def _arg(*args, **kwargs):
+    help='do not self-update West and the manifest before fetching project '
+         'data')
+
+# Flag for updating just the West source code repository, for 'selfupdate'
```
Sounds reasonable. I hope there's a nice and short way to get them in, without having to add a bunch of boilerplate.

Could move the single-use ones closer to the classes at least...

mbolivar replied:
> Sounds reasonable. I hope there's a nice and short way to get them in, without having to add a bunch of boilerplate.

Well, you an always call parser.add_argument(...) with the specification for the argument, no? IMO that actually removes boilerplate by eliminating the intermediate variable

ulfalizer replied:
Should work. Was thinking it was a default argument for a second.

ulfalizer replied:
Fixed

mbolivar commented:
> > Maybe auto-sync could be turned into an option. I suspect we might need a configuration file at some point anyway (something like git config), e.g. to remember the upstream branch for the West manifest.
> 
> I was just going to suggest this. A `~/.westconfig` would be perfect, but we need to think how to implement that on Windows. I think `%userprofile%\westconfig` (hidden) is probably the way to go.

I actually think it'd be nice to try to have per-installation configuration to start with -- so `TOP/west/westconfig.yml` -- thoughts?

mbolivar commented on src/bootstrap/main.py:
```diff
@@ -104,7 +97,7 @@ def repository_type(url):
 #
 
 
-def init(argv):
```
Makes it harder to use the bootstrapper as a library though...

ulfalizer replied:
If we ever need to use it as a library, we could add that then.

mbolivar replied:
1. Why do we need to delete it and add it back when we already have it?
2. One legitimate use for having this as parameters instead of assuming sys.argv is to call these functions from test cases to do a bootstrap in a tmpdir.

ulfalizer replied:
> Why do we need to delete it and add it back when we already have it?

Are we sure we'll add it back?

> One legitimate use for having this as parameters instead of assuming sys.argv is to call these functions from test cases to do a bootstrap in a tmpdir.

Could we wait until we're sure we'll use it?

Maybe there's a cleaner way to do it too, where application code doesn't have to change for tests. `pytest` might have stuff for mocking command-line arguments.

carlescufi replied:
Not a Python expert, but I do think not using the global sysargs is cleaner, passing them as a parameter instead.

ulfalizer replied:
@carlescufi 
Agreed in C at least, where you get the command-line arguments as arguments to `main()`. Storing them globally would be weird and unexpected there.

In Python, the command-line arguments are never implicitly passed to any startup function, and are accessed at `sys.argv` instead (have to `import sys` to get access to them). `sys.argv` immediately says "the command-line arguments" in idiomatic Python. `sys.argv` is kinda in the same class as `os.environ`, which holds the environment.

Think it comes down to whether the code ever needs to deal with anything but the command-line arguments. It would be simpler if it didn't.

carlescufi replied:
@ulfalizer fine, I do understand that. So they are a global system variable to start with. In that case I retract my comment since I did not know that.

mbolivar replied:
> > One legitimate use for having this as parameters instead of assuming sys.argv is to call these functions from test cases to do a bootstrap in a tmpdir.
> 
> Could we wait until we're sure we'll use it?

Well, I'm kind of heavily hinting that we *should* use it now in exactly this way ;)

ulfalizer replied:
@mbolivar 
Could you check if there's a way to avoid it while working on the bootstrap tests? In general, I think it's a bad idea to generalize code for testing if you can handle it within the tests itself. People won't know that it was added for the test.

mbolivar commented on src/west/__init__.py:
```diff
@@ -1,5 +0,0 @@
-# Copyright 2018 Open Source Foundries Limited.
```
> At the moment, `west` is just a small helper application at least. Maybe some of the packages will get reused later.

Right, that's what we're planning for, so we need to keep west importable. Only the `commands` subpackage contains the commands. Do you remember discussing this in the kickoff meetings?

ulfalizer replied:
I think you might be confusing the application with the packages it happens to be using. The application itself does not need to be importable.

mbolivar replied:
I'm not sure what distinction you are trying to make. Why should I not be able to import west?

ulfalizer replied:
Hmm... think I might've been confused by the directory structure.

Not saying we have to change it, but I think this directory might illustrate how I imagined it (at the top level of the repository, for example):

```
west.py           ("main" executable. Uses the West library.)
bootstrap.py      (could reuse some parts of West as well)
west/__init__.py  (the West package)
west/...
```

If the application is bundled inside the library at the moment, then I get it.

Sorry if you've already told me, but what other stuff besides `west.py` and `bootstrap.py` would need to use the West library?

ulfalizer replied:
Guess I'm confused about what we're making. Is it a library or an application? Will the library be used outside the application, so that e.g. some other application would be using the `runners/` classes?

carlescufi replied:
So from my point of view it should be both. An application that can be run directly from the command line, and also a library that can be imported and invoked directly by other scripts. So if `west` as a `sync` command, I need to be able both to type `west sync` from the command line but also to `import west` and then `west.sync()` from a script. That was my understanding at least.

ulfalizer replied:
@carlescufi 
Yeah, that'd make sense, if we need it to be a library as well.

Not sure what the public interface of the library is at the moment, and how much of it is generally useful. If we want to make it a proper library, we should define what the interface is, and should probably follow some conventions like prefixing non-public stuff with `_` (prevents it from showing up in documentation, and affects e.g. `from west import *`).

ulfalizer replied:
Will the library part ever be used by third-party applications not written within the Zephyr project? In that case, we need to at least do the following:

  1. Define the public API of the library (preferably prefixing internal stuff with `_`)

  2. Maintain API compatibility between versions

If the public API is "everything", then it's more of a grab bag of code than a library, and maintaining API compatibility will be extremely hard.

Going on personal experience, maintaining a library used by third parties is a ton more work than just having e.g. a shared codebase. If we don't need a library, we should definitely avoid it.

ulfalizer commented:
Fixed the bootstrap script. I renamed `WEST_TOPDIR` to `WEST_MARKER`. The old name confused me and made me think it was the path to the West top directory.

ulfalizer commented on src/bootstrap/main.py:
```diff
@@ -7,8 +7,7 @@
 import argparse
 import os
-import importlib
-import platform
```
Should be fixed now.

ulfalizer commented:
> I actually think it'd be nice to try to have per-installation configuration to start with -- so `TOP/west/westconfig.yml` -- thoughts?

I'd prefer some simple `.ini`-like format, like Git uses. Think `.yml` might be overkilling it.

This seems nice: https://docs.python.org/3/library/configparser.html

ulfalizer commented on src/west/commands/project.py:
```diff
@@ -235,6 +250,34 @@ def do_run(self, args, user_args):
+        super().__init__(
+            'selfupdate',
+            _wrap('''
+            Updates the West source code repository and/or the manifest
```
Changed the name to `update`.

mbolivar commented:
> I'd prefer some simple `.ini`-like format, like Git uses. Think YAML might be overkilling it.
> 
> This seems nice: https://docs.python.org/3/library/configparser.html

I've used configparser and suggested using it in this context before as well -- the counterargument I heard at the time is that Zephyr already use YAML for everything, so it's not adding new syntax, which configparser would be.

Kind of a judgement call.

ulfalizer commented:
> > I'd prefer some simple `.ini`-like format, like Git uses. Think YAML might be overkilling it.
> > This seems nice: https://docs.python.org/3/library/configparser.html
> 
> I've used configparser and suggested using it in this context before as well -- the counterargument I heard at the time is that Zephyr already use YAML for everything, so it's not adding new syntax, which configparser would be.
> 
> Kind of a judgement call.

I think it would be simpler for users to understand. YAML feels a bit "overstructured" for a simple configuration file for remembering the manifest branch and simple settings like that.

ulfalizer commented on src/bootstrap/main.py:
```diff
@@ -83,18 +82,12 @@ def is_west_dir(d):
 
 
 def clone(url, rev, dest):
-    def repository_type(url):
```
Will try to be better about it. I piggyback too much...

carlescufi commented:
@ulfalizer 

> I renamed WEST_TOPDIR to WEST_MARKER. 

+1 to that change, I also think it's clearer.

carlescufi commented:
@ulfalizer @mbolivar 
Regarding the config file format, I don't have a preference. Probably YAML makes strictly more sense, but then again sometimes it's good to be flexible. Git's format (configparser?) is perfectly fine for this kind of stuff.

mbolivar commented:
> **as long as it's documented** I think it makes sense to use the gold standard in `.config` file formats.

Makes sense.

I think we would want to be very careful about specifying a conservative subset of the full `configparser` syntax we expect, without the fancy bells and whistles of variable substitution etc., and documenting it carefully as you say if we go this route. If we ever have to re-implement that functionality in some language other than Python, KISS will help.

carlescufi commented:
@ulfalizer @mbolivar what are the outstanding issues with this PR so we can merge? I think I 
count:

- argv global vs local: optional IMHO
- select west branch: required (maybe cmd-line option for now, until we introduce the .config file)
- Keep `west` a package with an "API": required
- flake8: required? I see we use it in Zephyr too, so it would make sense here as well IMHO
- `.westconfig` file: optional

mbolivar commented:
The first 'init' seems to be fixed, but I'm still not able to build up a tree from scratch.

This time the problem is I can't fetch, even with an edited PYTHONPATH:

```
$ export PYTHONPATH=~/src/west/src
$ python ~/src/west/src/bootstrap/main.py init --wr selfupdate zephyrproject
Initializing in new directory zephyrproject
Cloning into 'zephyrproject/west/west'...
remote: Counting objects: 463, done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 463 (delta 16), reused 39 (delta 13), pack-reused 407
Receiving objects: 100% (463/463), 177.71 KiB | 1.52 MiB/s, done.
Resolving deltas: 100% (230/230), done.
Cloning into 'zephyrproject/west/manifest'...
remote: Counting objects: 13, done.
remote: Total 13 (delta 0), reused 0 (delta 0), pack-reused 13
Unpacking objects: 100% (13/13), done.
(west) dopey: /tmp mbolivar
$ cd zephyrproject/
$ python ~/src/west/src/bootstrap/main.py fetch
Traceback (most recent call last):
  File "/tmp/zephyrproject/west/west/src/west/main.py", line 23, in <module>
    from commands.project import ListProjects, Fetch, Pull, Rebase, Branch, \
ImportError: cannot import name 'SelfUpdate' from 'commands.project' (/tmp/zephyrproject/west/west/src/west/commands/project.py)
```

(My local tree `~/src/west` is on an up-to-date checkout of this branch).

Just to be sure, I did a `python setup.py install` in a clean virtualenv to get the `west` setuptools launcher script installed in the same way a user would end up with and tried it that way, same error:

```
$ west init --wr selfupdate zephyrproject-2
Initializing in new directory zephyrproject-2
Cloning into 'zephyrproject-2/west/west'...
remote: Counting objects: 463, done.
remote: Compressing objects: 100% (42/42), done.
remote: Total 463 (delta 16), reused 39 (delta 13), pack-reused 407
Receiving objects: 100% (463/463), 177.71 KiB | 1.33 MiB/s, done.
Resolving deltas: 100% (230/230), done.
Cloning into 'zephyrproject-2/west/manifest'...
remote: Counting objects: 13, done.
remote: Total 13 (delta 0), reused 0 (delta 0), pack-reused 13
Unpacking objects: 100% (13/13), done.
$ cd zephyrproject-2/
$ west fetch
Traceback (most recent call last):
  File "/tmp/zephyrproject-2/west/west/src/west/main.py", line 23, in <module>
    from commands.project import ListProjects, Fetch, Pull, Rebase, Branch, \
ImportError: cannot import name 'SelfUpdate' from 'commands.project' (/tmp/zephyrproject-2/west/west/src/west/commands/project.py)
```

I'd again like to bring up updating the test cases to do a real bootstrap instead of faking it to avoid these issues. Not sure if you had time to think about that one @ulfalizer.

mbolivar commented:
> @ulfalizer @mbolivar what are the outstanding issues with this PR so we can merge? I think I
> count:

Thank you for the summary!

> 
> * `argv` global vs param: optional IMHO

Since this change is unrelated to "self-update", I think it should be dropped from this PR since it's unrelated and still under discussion.

> * select `west` branch: required (maybe cmd-line option for now, until we introduce the `.westconfig` file)

I agree this is required, not sure how it is blocked by having a west config file.

I think the bootstrap script could keep track of the `--west-rev` (and `--manifest-rev`) arguments passed to `west init` in any number of ways (e.g. by running `git config west.rev SHA` or so when cloning the west repository at `init` time) that don't require a separate west config file.

> * Keep `west` a package with an "API": required

Agreed.

> * flake8: required? I see we use it in Zephyr too, so it would make sense here as well IMHO

I've sent a separate PR to clean up the flake8 patches. I'm fine with delaying that cleanup and making it mandatory until then. I'm against ignoring flake8 whenever we feel like it, because then its output gets too long to read, and it *does* catch real bugs (plus I am a fan of having a common code style! we can always configure out the parts we don't like).

> * `.westconfig` file: optional

Agreed.

I would add:

- should be possible to bootstrap a fresh west installation from scratch using this PR (using `west init` + `west fetch`) without errors -- preferably with regression testing, but I'm OK to defer that part
- from-scratch installation should update properly when its upstream manifest and west revisions defined at `init` time change (slightly different than your "select `west` branch: required" above)

ulfalizer commented:
recheck

ulfalizer commented:
Added a CI build fix for Python 3.4. Not sure why it didn't crop up before.

See the updated commit message.

ulfalizer commented:
> > * `argv` global vs param: optional IMHO
> 
> Since this change is unrelated to "self-update", I think it should be dropped from this PR since it's unrelated and still under discussion.

It's a simpler way to implement existing functionality, which is always a win in my book. Gotta make those little tweaks constantly to fight off complexity.

> > * Keep `west` a package with an "API": required
> 
> Agreed.

Is this a "don't change anything"? I'm wondering since we haven't defined a public API.

Are we making a set utilities with a shared codebase or a library? It will make a big difference re. what kind of API stability is needed for example.

I rank applications above shared codebases above libraries above frameworks personally. Super boring, straight-to-the-point, no-frills code is my favorite code. :)

> I agree this is required, not sure how it is blocked by havSuper boring, straight-to-the-point code is the best code. :)ing a west config file.
> 
> I think the bootstrap script could keep track of the `--west-rev` (and `--manifest-rev`) arguments passed to `west init` in any number of ways (e.g. by running `git config west.rev SHA` or so when cloning the west repository at `init` time) that don't require a separate west config file.

I don't think we should add things which we will immediately replace, but I'm not sure that this isn't what we'll go with either tbh.

I suspect a centralized configuration file might be easier to work with if we add more stuff, like making automatic self-updates configurable.

mbolivar commented:
> It's a simpler way to implement existing functionality, which is always a win in my book. > Gotta make those little tweaks constantly to fight off complexity.

Piggybacking this change into the same commit that's adding the self-update feature (which we are eager to merge) kind of feels like forcing it through as a cleanup when it's not that simple, TBH. Especially since it seems to be a part of a larger conversation about development style and what we're doing here.

You may prefer applications to libraries to frameworks, but it is far from clear to me that we're just developing an application. Quite the opposite actually.

And either way, it's definitely muddying the waters to have all of this in one big patch.

I'm feeling increasingly like I should keep my -1 on until the PR is just about self-update, or (if it keeps the other portions that are unrelated) until we've had enough time to discuss all the other issues and come to consensus.

ulfalizer commented:
Looks like you can do this with `pytest`:

```python
with patch('sys.argv', ['foo', 'bar']):
    main()
```

Could do this to make it as easy to use from tests as if `main()` took `argv`:

```python
def run_with_args(args):
    with patch('sys.argv', ['west'] + args):
        main()

run_with_args(['checkout', 'foo'])
```

(With an extra `split()` in `run_with_args()`, it could become `run_with_args('checkout foo')` too.)

That avoids adding stuff to the application code that would only be used by tests. With a parameter, you probably ought to test that the `sys.argv` fallback works as well, which kinda defeats the purpose.

I'll try to piggyback less in the future, but this is a pretty harmless simplification.

tejlmand commented:
@ulfalizer Great work so far :)

But let me give my comments to the discussion:

> > > > - argv global vs param: optional IMHO

> > > Since this change is unrelated to "self-update", I think it should be dropped from this PR since it's unrelated and still under discussion.

> > It's a simpler way to implement existing functionality, which is always a win in my book. Gotta make those little tweaks constantly to fight off complexity.

> Piggybacking this change into the same commit that's adding the self-update feature (which we are eager to merge) kind of feels like forcing it through as a cleanup when it's not that simple, TBH. Especially since it seems to be a part of a larger conversation about development style and what we're doing here.

Agree with @mbolivar  here.
I have already been facing this in Zephyr related PR's. Piggybacking too much myself :(
It feels natural when changing the code to do small clean-up on the same PR, but it makes it much harder to review, and keep the focus / discussion on the important part.

Just look at this PR and the spawned discussion. Most of it is not addressing the important self-updating feature, but other things.

> - select west branch: required (maybe cmd-line option for now, until we introduce the .westconfig file)

It appears to me we already have this with the `west init --wr <branch/tag>` possibility.
The `.westconfig` should be described and discussed as an enhancement in it's own issue.

> > > - Keep west a package with an "API": required

> > Agreed.

> Is this a "don't change anything"? I'm wondering since we haven't defined a public API.

I guess no, but it is, don't change it in this PR.
Then this discussion of changing the API can be taken separately.
So I agree with @mbolivar  and @carlescufi here

> Are we making a set of utilities with a shared codebase or a library? It will make a big difference re. what kind of API stability is needed for example.
>
> I rank applications above shared codebases above libraries above frameworks personally. Super boring,  straight-to-the-point, no-frills code is my favorite code. :)

I strongly disagree here.
When doing bigger works in python (i.e. not a small shell script), I always think of the core functionality in an OO way. A module with clear API, and then the app as a pure and thin command line parser.
The advantages of this is:
- Reuse the core functionality by other scripts, as @mbolivar also shows
- It can make it easier to integrate into build-servers, where you don't need a command line.
  And using semantic versioning, it is clear when the API changes, https://semver.org/
- Makes it easier by other, e.g. Oticon, to reuse core functionality and place their extended needs in their own code, thus not having to add their changes in the `west` repo.
- Someone can easily create a `west-gui` tool using the same `west` module, but providing a gui interface with pull down menus, as example.

So baseline is imho: don't remove this possibility in this PR.
If you want to discuss that, then make a separate issue / PR.

> > - flake8: required? I see we use it in Zephyr too, so it would make sense here as well IMHO

> I've sent a separate PR to clean up the flake8 patches.

@mbolivar Nice, so let's discuss this topic there

> .westconfig file: optional

Agree, see also comment above about `.westconfig`.

 
final note: Rest of work is great ;)

ulfalizer commented:
@tejlmand 
I'm the Kconfiglib guy, so I'm not opposed to libraries. I use semantic versioning for it. Unlike some other projects which seem to have adopted it for hipster points, I increase the major version even for tiny API behavior changes. :)

> When doing bigger works in python (i.e. not a small shell script), I always think of the core functionality in an OO way. A module with clear API, and then the app as a pure and thin command line parser.

That sounds like plain old modular code to me though. By library, I mean something usable by third-party code, preferably with a stable API.

Here's what I mean by shared codebase vs. library:

 - The code in the Linux kernel is a shared codebase. There's no guarantee that APIs won't change, and whenever they do, all clients are updated at the same time. Provided you control all the clients, this is the simplest model.

 - The user-space API is more like a library, and is guaranteed to be stable. If you deliberately break it, Linus screams at you.

What I'm opposed to is making a library when all you need is an application or a few applications with a shared codebase. It doesn't make things fancier to me, just messier. I'm a strong believer in always going for the simplest thing you can think of (reducing boilerplate and visual clutter is super big for me), until you're absolutely sure that you need more than that.

Rambling aside, if we need to make West stuff importable outside of West, we ought to have a clearer separation between the application parts and the reusable parts.

@mbolivar 
Is it just the runners that are meant to be reusable outside of the West codebase? IIRC you said that yesterday on IRC.

In that case, maybe we could have a structure like the following one:

```
bootstrap.py
westcmd.py
commands/...
west/runners/...
```

`westcmd.py` would import `commands`, and `west.runners` to get access to the runners. Other stuff could import `west` as well.

I would just go for `runners/` and change all the clients later if it turns out that more stuff is needed, but pretty sure that would get shot down, so I pre-give-up. ;)

@tejlmand Any thoughts on the above? Think it might be something similar to what you had in mind.

tejlmand commented:
@ulfalizer 
> I'm the Kconfiglib guy, so I'm not opposed to libraries. I use semantic versioning for it. Unlike some other projects which seem to have adopted it for hipster points, I increase the major version even for tiny API behavior changes. :)

Yes, I know.
And I think everyone here know semantic versioning.
Main point was just to list the advatages.
After all, I think we are heading in same direction, just looking at it from different perspectives.

> > When doing bigger works in python (i.e. not a small shell script), I always think of the core functionality in an OO way. A module with clear API, and then the app as a pure and thin command line parser.

> That sounds like plain old modular code to me. By library, I mean something usable by third-party code, preferably with a stable API.

Yes, the way you structure code internally doesn't make it more or less a library.
The library is more something seen from outside and with stable API.
Think we agree there, just wording it differently ;)

But as it is early days for `west` and version < v1.0.0, we can change all that we want.

> Here's what I mean by shared codebase vs. library:
>
 >   The code in the Linux kernel is a shared codebase. There's no guarantee that APIs won't change, and whenever they do, all clients are updated at the same time. Provided you control all the clients, this is the simplest model.
>
>    The user-space API is more like a library, and is guaranteed to be stable. If you deliberately break it, Linus screams at you.

Yes, I think we agree here. 
The internals of the library / shared codebase can change irregardless of the API versioning.
The main point here is: "Provided you control all the clients"
Do we control all clients in future ?
I have a feeling much of this discussion is our expectation to `west` and how it may be used.
Therefore I have created: #38 as this is indeed an important discussion.

> What I'm opposed to is making a library when all you need is an application or a few applications with a shared codebase. It doesn't make things fancier to me, just messier. I'm a strong believer in always going for the simplest thing you can think of (reducing boilerplate and visual clutter is super big for me), until you're absolutely sure that you need more than that.

I can only agree to not over-engineer things from the start.
A very sane principle.
But as pointed out already, several use-cases has been described, where `west` as a library makes sense.
And in that case, it makes sense to take that into consideration from the start.

> @mbolivar
> Is it just the runners that are meant to be reusable outside of the West codebase? IIRC you said that yesterday on IRC.
> 
> In that case, maybe we could have a structure like the following one:
>
> bootstrap.py
> westcmd.py
> commands/...
> west/runners/...
> 
> westcmd.py would import commands, and west.runners to get access to the runners. Other stuff could import west as well.
> 

Can we agree to continue this in: #38 ?

> I would just go for runners/ and change all the clients later if it turns out that more stuff is needed, but pretty sure that would get shot down, so I pre-give-up. ;)

No, don't pre-give-up.
I think it is a very important discussion and you definitely have valuable experience here.
I always listen, even though I may not always agree ;-)

> @tejlmand Any thoughts on the above? Think it might be something similar to what you had in mind.

Seems reasonable, although, as you also points out, we should not namespace too deep.
so just have:
```
from west import commands OR  from west import WestCommands
from west import runners OR from west import WestRunners
```
is still nice to be able to.
As I see it, the bastard in all this is the bootstrap, which is a `west` related thing, but not west itself.

ulfalizer commented:
PR updated.

The CI fix is needed on 3.4 (or was yesterday, anyway... wonder if some CI change triggered it).

I also attached a small obvious bug fix for `find_west_topdir()` that broke `west init <directory>`. Not sure it's all that related in retrospect though.

ulfalizer commented:
> Yes, the way you structure code internally doesn't make it more or less a library.
> The library is more something seen from outside and with stable API.
> Think we agree there, just wording it differently ;)

Ah, sorry, misunderstood. Thought you meant that modular code == library.

Alright, will move over to https://github.com/zephyrproject-rtos/west/issues/38 for anything else.
